### ISSUE TITLE
Dynamically updating the database schema

### ISSUE THREAD

**Issue by isedwards** [opencdms/Climsoft#304]
In a few places in the code, the database schema is currenlty altered dynamically at run time
(e.g. [frmDataForms](https://github.com/opencdms/Climsoft/blob/98e2c644c1e16c0642269419be85b5dac982b589/ClimsoftVer4/ClimsoftVer4/frmDataForms.vb#L19-L24))

```
ALTER TABLE `data_forms`
    ADD COLUMN `entry_mode` TINYINT(2) NOT NULL DEFAULT '0' AFTER `sequencer`;
```

@smachua - Do you think this code should be moved to the splash screen instead (so that it is attempted immediately when the user first logs in?).  If the user does not have admin priviledges on the database then it would be good to warn them early on that the required update has failed.

Also, do you have:
a) the script that was used to create the Climsoft 4 database in Meteo Rwanda?
b) the alter commands for all of the changes since?

With these, we should be able to automatically update the Meteo database when they receive the next version of the software.  We can use the same process in the future for all organisations to automatically update when newer versions of the software are installed.

**Comment by mhabimana** [comment]
@isedwards : Good to see that the merging. I am now able to clone and run the project without errors. However, victor's contribution is still missing.

**Comment by iedwards** [comment]
Do you think @vralenkoane will submit his latest version as a pull request so that it can be merged automatically, or should we use the version that you both have?

**Comment by mhabimana** [comment]
@isedwards : If you replace the current files for frmPlotter and formProductsSelectCriteria  with the one in the PR Samuel sent earlier,his contribution will be part of the project.

**Comment by smachua** [comment]
Now that we have a merged code in Github wouldn't be better to start making commits to a single repository rather the separate two? That is my thinking. What do others think?

**Comment by mhabimana** [comment]
@smachua : I think it would be great if you elaborate more on some of the issues you shared with us  yesterday which may affect the climsoft cdms if the auto increment new field Id added in the database schema tabled by @isedwards are not removed. If I remember well, one of the things you mentioned is that this will affects the speed of data processing of climsoft CDMS .

**Comment by isedwards** [comment]
It would be good to know more. The technologies (like MariaDB) are a lot more advanced now than the original database that was used in the first Climsoft version.

Most of the modern frameworks make the assumption that there will be a single-column primary key in order to increase the speed and efficiency.  I'd be surprised if adding an auto incrementing field to a modern database caused any decrease in performance.

**Comment by isedwards** [comment]
The composite key prevents duplicates from being created, but we can achieve the same restriction by having a UNIQUE constraint on the columns that formed the composite key.

**Comment by maxwellfundi** [comment]
@smachua @mhabimana I can confirm now that we will be working on the openCDMS repository @Patowhiz  this is now discussed and decided so we need to start working from there now. 

@isedwards  From what i got from @smachua yesterday was that adding this new column too added issues to his existing code which now does not work properly. (not sure of the finer details of this)

**Comment by mhabimana** [comment]
@maxwellfundi : well noted.

**Comment by isedwards** [comment]
If Samuel's code references the fields by column number, instead of by name, then it would cause a problem. @smachua is this the case, or is there another issue?

If this is the problem then it could be avoided by adding `id` as the final column, then everyone's code would work.

**Comment by smachua** [comment]
The composite key not only prevents duplicates but useful in many functions such as referencing a record for retrieval and updating using input values. Values in the id fields will be auto generated and  not simply used in such functions. Yes, some operations have been affected such as AWS realtime, TDCF encoding and paperarchiving. The affected tables can be modified but with no much benefits.

From the past experiences the auto key will add more data into the tables especially the observationfinal where records run into thousands of millions in some Met Services. You may consult @Steve-Palmer on the discussions we had until we decided to do away with it.

Since the need for the id field is only in uploading of data from the new forms I recommend that the code for doing that be modified otherwise and we do away with the requirement for that field.

**Comment by Steve-Palmer** [comment]
@smachua is correct on the reasons for using a composite primary key, especially on the main observations table. We need to be absolutely sure that there are no duplicate records for an observation. Also the observations table will be huge, with many very small records. So adding an id attribute to each of these records would add significantly to the storage requirements. 

It would be possible to add an id attribute- but this would need very very strict rules to ensure that the id never allowed duplicate records. It may have other consequences.

Retrieval of climate data essentially falls into two types. 
(A) A number of elements for one or more stations over a short period of time e.g. a monthly summary or a monthly map. This type is most common for recent data, within the last two years.
(B) one element for one or a small number of stations over a very long time, e.g. 30-year means and extremes. 
These two types are essentially orthogonal in terms of the exact arrangement of the data in the database. Using an id number may well constrain this arrangement, so that one type of analysis is efficient and the other not. In many RDBMS it is possible to specify the storage order. There may be operational efficiencies by partitioning the observation tables so that the recent data is optimised for type A queries, and older data is optimised for type B.

The ideal process is that an observation is made and then entered into the database, along with the date/time location and type of observation. At this stage it is the only record containing that observation. Subsequently that observation may be corrected by QC action (there is also a potential change where a more accurate value may be received e.g. where temperature in whole degC received in a METAR may be replaced by temperature in 0.1degC from a Synoptic report). There may be multiple stages of QC correction. However, the business rules which are encapsulated in WMO-1131 only require the database to hold the "original received" and "latest QC'd" values of an observation element where these are different (along with flags indicating the status of QC checks applied, and the reasons for any QC change made). Historically some CDMS have included full roll-back of QC changes, but in practice this facility is never used. The Climsoft data model had a "Version number" attribute from the beginning, ( where Version=1 is the latest QC'd value, and Version=0 is the "original received"; Version=0 only exists when a QC change has been made.) but this was not implemented in earlier releases.

It is very rare that a QC change will be made to an observation value when it is more than two years old, so there may be operational avantages in archiving the older data so it is read-only, and requires a higher level of authority than "QC Operator" - this feeds into some of the climate change deniers allegations of fiddling the data. Again, partitioning the database may facilitate this.

My feeling is that for the main observations table, I would much rather keep the composite primary key, and pay the penalty in reduced performance. I suspect that modern RDBMS would do this optimisation internally. This could be revisited later, perhaps as part of the requirements for Climsoft V5 or OpenCDMS.

**Comment by isedwards** [comment]
Good, this all sounds sensible.

@Steve-Palmer, @smachua are you happy with having single-column primary keys in all of the tables except for the large open tables (`observationinitial` and `observationfinal`)?  We can then create workarounds for accessing the observation tables, but continue to use all of the code that AMI have written in recent months.

I see other benefits to having the single-column ids in the majority of tables for future versions of Climsoft/OpenCDMS -- esp. if there is the possibility of using web frameworks in the future.

We can still keep composite keys (also called "unique indexes") in any tables that need them. This will prevent duplicates etc., they would just no longer be the primary key...

**Comment by isedwards** [comment]
Ok, agreed.

@maxwellfundi, @Patowhiz - can you make sure that you have the latest version of the database from @smachua with all of his changes (and can you let us know [which tables you lose access to](https://github.com/africanmathsinitiative/Climsoft/issues/75) - is it just `form_monthly`, `observationschedule` and `observationinitial`)?

@volloholic, @dannyparsons - we'll need to workaround any issues with Entity Framework by writing [SQL code](https://github.com/opencdms/Climsoft/blob/965603b372596899e0262c84d6cf66d8e4cdc476/ClimsoftVer4/ClimsoftVer4/formMonthly.vb#L830-L832).

**Comment by maxwellfundi** [comment]
@smachua Kindly l attach here the latest database version so that we can do a check over the weekend. We should do this quickly @patowhiz to make sure we give feedback fast for good progress to happen

**Comment by isedwards** [comment]
Also - I need to create a branch with an updated version of the entity model for testing with

**Comment by mhabimana** [comment]
@maxwellfundi : re-your message above,which database are you refer to? may be you forgot to add the link.

**Comment by maxwellfundi** [comment]
@mhabimana  I was suggesting that @smachua to send us his latest updated database.

**Comment by Fortunata** [comment]
Any meteorological variable is identified at a point or station from which it was observed and also which was the time of  its observation which makes these variables known as (values in entity tables) unique hence the composite keys. In the case of the station tables, the station_ids were made the primary keys but they are instances when those numbers are not sequential hence the need of the auto_id as a primary key. Need more debate

**Comment by mhabimana** [comment]
@smachua kindly share with the team the latest climsoft database schema.

**Comment by smachua** [comment]
@mhabimana , @maxwellfundi , @iedwards , @Patowhiz find in the attachment the script file for both the current test and operational database schemas. Let me know if anything fails to work as expected.
[mariadb_climsoft_db_v4_all.txt](https://github.com/opencdms/Climsoft/files/2241419/mariadb_climsoft_db_v4_all.txt)

**Comment by maxwellfundi** [comment]
Thank you @smachua

**Comment by Patowhiz** [comment]
@smachua according to the screenshot, I expected element 185 (Evaporation) to have a corresponding Val_Elem185 column in the database table. Currently we only have its corresponding Flag185. Even after running the script you attached above. Have a look at it. 

![downloaded](https://user-images.githubusercontent.com/21179007/43516887-90e2634c-958f-11e8-8801-d0ddc41a3ff3.png)

**Comment by Patowhiz** [comment]
Even column for **element 99 (Grass Min Temp)** is missing. I think this will be **Val_Elem099**. We currently only have **Flag099**

**Comment by Patowhiz** [comment]
Sorry I have seen them. They are all there

**Comment by Patowhiz** [comment]
@mhabimana  @smachua in the above screenshot Geopotential is labelled as **element 185**. That means **Evaporation** and **Geopotential** have the same element code? Should this be **196**?

**Comment by Patowhiz** [comment]
@smachua @isedwards @maxwellfundi @mhabimana  for form synoptic to work, we need the columns  Flag169, Flag170, Flag171 both in the database and the EF model . Their corresponding Val_Elem169, Val_Elem170, Val_Elem171 are already there. 

I have fixed the form, when the above is fixed then the form can be tested.

**Comment by mhabimana** [comment]
@Patowhiz @smachua @isedwards @maxwellfundi Attached is the updated form_synoptic_2_ra1: I have added the above mentioned 3 flags:
[form_synoptic_2_ra1.txt](https://github.com/opencdms/Climsoft/files/2249702/form_synoptic_2_ra1.txt)

**Comment by Patowhiz** [comment]
@mhabimana thank you. @isedwards I hope we can have the updated model soon, as we conclude the upload functionality.

**Comment by mhabimana** [comment]
@Patowhiz @smachua @isedwards @maxwellfundi: the reverted EF is now working well.What is remaining is the upload button on all the data entry forms,update the EF model with the above form_synoptic_2_ra1. Unless I forgot some thing. However,I am able to switch the database connection at login (from operation database to the test database), but not internally.Here I mean under Administration menu,database connection tab.

**Comment by isedwards** [comment]
@mhabimana - `form_synoptic_2_ra1.txt` currently defines a table that does not have a primary key (only a composite index).  I will change this to a composite primary key, like we did with `form_monthly` and then update the model.  I will also attach a new version of the `form_synoptic_2_ra1.txt` script.

**Comment by mhabimana** [comment]
@isedwards Thanks I can see in the reverted database schema the PK is applied at stationid,year(yyyy),month(mm) and hour(hh)
![image](https://user-images.githubusercontent.com/10340170/43530853-38f735cc-95b7-11e8-8129-afea1334398c.png)

**Comment by isedwards** [comment]
```
ALTER TABLE `form_synoptic_2_ra1`
	DROP INDEX `stationId_yyyy_mm_dd_hh`,
	ADD PRIMARY KEY (`stationId`, `yyyy`, `mm`, `dd`, `hh`);
```

**Comment by mhabimana** [comment]
@isedwards Thanks

**Comment by smachua** [comment]
@isedwards you're right. Any composite key in any table can be made Primary key. I may have omitted some.

**Comment by mhabimana** [comment]
@Patowhiz @smachua : Grass Min box(formSynop_2_ra1 is not active on all the hours.I was expecting this cell to be active at least at 0600 UTC. I am not sure if this is an omission or the way this was programme.

**Comment by smachua** [comment]
@mhabimana and @Patowhiz Grass min and Min temp entries should be enabled once daily according to the time set in General Settings.

**Comment by Patowhiz** [comment]
@mhabimana have you checked on the settings? As outlined by @smachua

**Comment by mhabimana** [comment]
@Patowhiz  @smachua Yes,I am now OK.Grass Min is active from April to September only ,this is set in the general settings.

**Comment by Patowhiz** [comment]
I'm working on the progress bar for the upload. In the meantime I'm to add any improvement that is urgent

**Comment by mhabimana** [comment]
@Patowhiz  Thanks, Not sure if there is any remaining task as of now apart the progress bar @isedwards  and @smachua  can let us know.

**Comment by mhabimana** [comment]
@Patowhiz @maxwellfundi @smachua form_synoptic_2_ra1 is not active from September to December for all hours,not sure if this is they way it was set.

**Comment by mhabimana** [comment]
@Patowhiz @maxwellfundi @smachua Add New Button on the Form Monthly is not giving an empty form.
